addons/mail/models/mail_activity.py

```diff
@@ -355,16 +357,25 @@ def write(self, values):
                         self.env['bus.bus'].sendone(
                             (self._cr.dbname, 'res.partner', partner.id),
                             {'type': 'activity_updated', 'activity_deleted': True})
+        if values.get('activity_type_id'):
+            for activity in self.filtered(lambda x: x.res_id and x.res_model):
+                self.env[activity.res_model].browse(activity.res_id)._set_activity_exception_type()
         return res
 
     @api.multi
     def unlink(self):
+        record_values = []
```

Comment:
record_values could be renamed + self.env[record[0]].browse(record[1]) is no very elegant.
you could immediately add the browse in the list